File: leetcode-solutions/ruby/111. Minimum Depth of Binary Tree.rb
# https://leetcode.com/problems/minimum-depth-of-binary-tree/

def min_depth(root)
    return 0 unless root
 
    if root.left && root.right 
         return [min_depth(root.left), min_depth(root.right)].min + 1
    elsif root.left
        return min_depth(root.left) + 1
    else
        return min_depth(root.right) + 1
    end
 end

File: leetcode-solutions/python/111. Minimum Depth of Binary Tree.py
# https://leetcode.com/problems/minimum-depth-of-binary-tree/

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def minDepth(self, root: Optional[TreeNode]) -> int:
        if root is None:
            return 0
        s = [root]
        depth = 0
        while s:
            depth += 1
            size = len(s)
            temp = []
            for _ in range(size):
                node = s.pop()
                if node.left is None and node.right is None:
                    return depth
                if node.left:
                    temp.append(node.left)
                if node.right:
                    temp.append(node.right)
            s = temp
        return depth
